api/utils/keys/yubikey.go

```diff
@@ -586,29 +657,71 @@ func (y *YubiKey) checkOrSetPIN(ctx context.Context) error {
 		return trace.Wrap(err)
 	}
 
-	yk, err := y.open()
-	if err != nil {
-		return trace.Wrap(err)
-	}
-	defer yk.Close()
-
 	switch pin {
 	case piv.DefaultPIN:
 		fmt.Fprintf(os.Stderr, "The default PIN %q is not supported.\n", piv.DefaultPIN)
 		fallthrough
 	case "":
-		if pin, err = setPINAndPUKFromDefault(ctx, yk); err != nil {
+		if pin, err = y.setPINAndPUKFromDefault(ctx); err != nil {
 			return trace.Wrap(err)
 		}
 	}
 
-	return trace.Wrap(yk.VerifyPIN(pin))
+	return trace.Wrap(y.verifyPIN(pin))
 }
 
-// open a connection to YubiKey PIV module. The returned connection should be closed once
-// it's been used. The YubiKey PIV module itself takes some additional time to handle closed
+type sharedPIVConnection struct {
+	// card is a reader name used to find and connect to this yubiKey.
+	// This value may change between OS's, or with other system changes.
+	card string
+
+	// conn is the shared PIV connection.
+	conn              *piv.YubiKey
+	mu                sync.Mutex
+	activeConnections int
+}
+
+// connect establishes a connection to a YubiKey PIV module and returns a release function.
+// The release function should be called to properly close the shared connection.
+// The connection is not immediately terminated, allowing other callers to
+// use it before it's released.
+// The YubiKey PIV module itself takes some additional time to handle closed
 // connections, so we use a retry loop to give the PIV module time to close prior connections.
-func (y *YubiKey) open() (yk *piv.YubiKey, err error) {
+func (c *sharedPIVConnection) connect() (func(), error) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	release := func() {
+		c.mu.Lock()
+		defer c.mu.Unlock()
+		if c.activeConnections > 0 {
+			c.activeConnections--
+		}
+
+		// If this is the last active connection, close after a delay,
+		// giving other callers a chance to claim the connection first.
+		if c.activeConnections == 0 {
+			go func() {
+				time.Sleep(releaseConnectionDelay)
+
+				c.mu.Lock()
+				defer c.mu.Unlock()
+
+				// Close the shared connection if there are still no new active connections,
+				// and the connection hasn't been yet closed by another goroutine.
+				if c.activeConnections == 0 && c.conn != nil {
+					c.conn.Close()
+					c.conn = nil
+				}
+			}()
+		}
+	}
```

Comment:
One fix and a couple nits.

fix: We should actually sleep from the start, before decrementing active connections. Otherwise, we could have the following situation: client connects, client disconnects. 4 seconds later, second client connects, and disconnects. 1 second later, the first client closes the connection. Instead, we should wait 5 seconds from the second client disconnect before closing. So we need to run the goroutine+sleep before decrementing the active connections.

nit: Now that we are decrementing and closing under the same lock, we can safely remove the `c.conn != nil` check, this should never be the case anymore.

nit2: We also don't need the `if c.activeConnections > 0` check. Every `connect` should be accompanied by `release`, so active connections will always be > 0 in this check, before it's decremented to its minimum value 0.

IMO the nits make it easier to implement bugs in the future, I'd rather the system fail loudly with faulty code (panic) so we fix it promptly.

```suggestion
	release := func() {
		go func() {
			time.Sleep(releaseConnectionDelay)
			
			c.mu.Lock()
			defer c.mu.Unlock()
			
			c.activeConnections--
			if c.activeConnections == 0 {
				c.conn.Close()
				c.conn = nil
			}
		}
	}
```

Reply:
>Otherwise, we could have the following situation: client connects, client disconnects. 4 seconds later, second client connects, and disconnects. 1 second later, the first client closes the connection. Instead, we should wait 5 seconds from the second client disconnect before closing.

Yeah, I thought about it, but for some reason it seemed to me that the correct behavior is to close the connection only after waiting for the first client to release 🤦. But I agree, it makes much more sense to extend the duration with each connected client.
Fix applied.